### PR TITLE
oci: do not use conmon for exec sync

### PR DESCRIPTION
**Pull request by haircommander**
#### What type of PR is this?

/kind design
#### What this PR does / why we need it:
It is not really needed, and causes unnecessary overhead. Plus, we drop an unkillable child from the mix
allowing cri-o to keep track of its children better.

Signed-off-by: Peter Hunt <pehunt@redhat.com>
#### Does this PR introduce a user-facing change?

```release-note
ExecSync requests now don't use conmon, instead calling the runtime directly, which reduces overhead.
```

### CODE REVIEW

File: internal/oci/runtime_oci.go [hunk 1]
```diff
@@ -484,85 +359,44 @@ func (r *runtimeOCI) ExecSyncContainer(ctx context.Context, c *Container, comman
...
-	stdoutBytes, stderrBytes := parseLog(ctx, logBytes)
 	return &ExecSyncResponse{
-		Stdout:   stdoutBytes,
-		Stderr:   stderrBytes,
-		ExitCode: ec.ExitCode,
-	}, nil
+		Stdout:   stdout,
+		Stderr:   stderr,
+		ExitCode: exitCode,
+	}, waitErr
```
[the remark this answers is not in the document]
Comment (mrunalp) [reply internal/oci/runtime_oci.go:389]: Please check if the kubelet will tolerate this change.

[the remark this answers is not in the document]
Comment (haircommander) [reply internal/oci/runtime_oci.go:389]: there's something awry with the code. I am working on getting critests passing, and that will likely involve changing this block

File: internal/oci/runtime_oci.go [hunk 2]
```diff
@@ -456,26 +339,18 @@ func (r *runtimeOCI) ExecSyncContainer(ctx context.Context, c *Container, comman
...
 	defer os.RemoveAll(processFile)
 
-	args = append(args,
-		"--exec-process-spec", processFile,
-		"--runtime-arg", fmt.Sprintf("%s=%s", rootFlag, r.root))
-
-	cmd := exec.Command(r.config.Conmon, args...) // nolint: gosec
+	execCtx := ctx
+	if timeout > 0 {
+		execCtx, _ = context.WithTimeout(ctx, time.Second*time.Duration(timeout))
```
Comment (saschagrunert) [inline internal/oci/runtime_oci.go:344]: We should probably cancel this new context once the exec is done.

Comment (haircommander) [reply internal/oci/runtime_oci.go]: good point, fixed!

File: internal/oci/runtime_oci.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (haircommander) [reply internal/oci/runtime_oci.go]: I just spent nearly a whole day looking for
```
        // gather exit code from waitErr
        exitCode := int32(0)
        if waitErr != nil {
-               if exitError, ok := err.(*exec.ExitError); ok {
+               if exitError, ok := waitErr.(*exec.ExitError); ok {
                        exitCode = int32(exitError.ExitCode())
                }
        }
```
:cry:

**Comment by haircommander** [comment]
goodness this is a rabbit hole
we needed to handle the case where the command times out but the container process still runs. since the container process isn't a child of the runtime process, we can't kill the exec.Command struct, but rather find the process with pidFile and kill that (what conmon does).

I think this passes critest now, hard to tell because I keep getting rate limited. no idea what's up with the openshift jenkins tests either...

File: internal/oci/runtime_oci.go [hunk 3]
```diff
@@ -393,176 +320,138 @@ func (r *runtimeOCI) ExecContainer(ctx context.Context, c *Container, cmd []stri
...
+	// TODO(haircommander): There does exist a race that we could time out before the runtime actually creates the file.
+	// Should we do inotify on this file to ensure it exists? Is that overkill?
+	ctrPid, ctrPgid, err := pidAndpgidFromFile(pidFile)
 	if err != nil {
-		return nil, &ExecSyncError{
-			Stdout:   stdoutBuf,
-			Stderr:   stderrBuf,
-			ExitCode: -1,
-			Err:      err,
+		log.Errorf(ctx, "failed to get pid (%d) or pgid (%d) from file %s: %v", ctrPid, ctrPgid, pidFile, err)
```
Comment (saschagrunert) [inline internal/oci/runtime_oci.go:426]:
```suggestion
		log.Errorf(ctx, "Failed to get pid (%d) or pgid (%d) from file %s: %v", ctrPid, ctrPgid, pidFile, err)
		return
```

File: internal/oci/runtime_oci.go [hunk 4]
```diff
@@ -393,176 +320,138 @@ func (r *runtimeOCI) ExecContainer(ctx context.Context, c *Container, cmd []stri
...
+		// First attempt to kill the container group
+		if err := syscall.Kill(-ctrPgid, syscall.SIGKILL); err != nil {
+			log.Errorf(ctx, "Failed to kill process group after timeout: %v", err)
+		}
+	} else if ctrPid > 0 {
+		// If that fails, kill the container PID itself
+		if err := syscall.Kill(ctrPid, syscall.SIGKILL); err != nil {
+			log.Errorf(ctx, "Failed to kill process after timeout: %v", err)
 		}
 	}
```
Comment (saschagrunert) [inline internal/oci/runtime_oci.go:440]:
```suggestion
```

**Review by saschagrunert — APPROVED** [review]
Just two nits, otherwise LGTM

File: internal/oci/runtime_oci.go (hunk 3 above)
Comment (haircommander) [reply internal/oci/runtime_oci.go]: fixed!

File: internal/oci/runtime_oci.go (hunk 4 above)
Comment (haircommander) [reply internal/oci/runtime_oci.go]: fixed!

File: internal/oci/runtime_oci.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (saschagrunert) [reply internal/oci/runtime_oci.go]: Oh, nice catch

**Review by saschagrunert — APPROVED** [review]
/lgtm
/hold
@mrunalp @haircommander feel free to lift the hold when ready.

**Comment by saschagrunert** [comment]
Backported manually in https://github.com/cri-o/cri-o/pull/4953 and https://github.com/cri-o/cri-o/pull/4954

**Pull request merged**